source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,272 @@
+              @Override
+              public void set(Object value) {
+                   setFlying(((Boolean) value).booleanValue());
+              }
+          });
+  }
+
+That's it. The ``DataManipulator`` should be done now.
+
+2. ImmutableDataManipulator
+===========================
+
+Just repeat the above mentioned steps for the ``ImmutableDataManipulator``.
+
+Now, after the ``Manipulators`` are done we have to register the ``KEY`` and finally implement the
+``DataProcessor`` and ``ValueProcessor``.
+
+3. Register the Key in the KeyRegistry
+======================================
+
+The next step is to register your ``KEYS`` in the ``KeyRegistry``:
+
+.. code-block:: java
+
+ public static void registerKeys() {
+      keyMap.put("is_flying", makeSingleKey(Boolean.class, Value.class, of("IsFlying")));
+ }
+
+
+4. Implement the DataProcessor
+==============================
+
+Next up is the ``DataProcessor``. A ``DataProcessor`` implements your ``DataManipulator`` to work semi-natively with
+Minecraft's objects. Unfortunately, since it's always easier to just directly implement all of the processing of a
+particular set of data in one place, we have ``DataProcessors`` to do just that.
+Currently, a ``DataProcessor`` has the following methods:
+
+.. code-block:: java
+
+ public interface DataProcessor<M extends DataManipulator<M, I>, I extends ImmutableDataManipulator<I, M>> {
+
+    int getPriority();
+
+    boolean supports(DataHolder dataHolder);
+
+    Optional<M> from(DataHolder dataHolder);
+
+    Optional<M> createFrom(DataHolder dataHolder);
+
+    Optional<M> fill(DataHolder dataHolder, M manipulator, MergeFunction overlap);
+
+    Optional<M> fill(DataContainer container, M m);
+
+    DataTransactionResult set(DataHolder dataHolder, M manipulator, MergeFunction function);
+
+    Optional<I> with(Key<? extends BaseValue<?>> key, Object value, I immutable);
+
+    DataTransactionResult remove(DataHolder dataHolder);
+ }
+
+getPriority() method
+--------------------
+
+The first thing that we notice is a ``getPriority()`` method. This is used for multi-processor registration, or in more
+plain terms: A multi-processor registration system for mods to provide their own compatibility processors for their
+custom entity/itemstack/tileentity/blockstate types. The higher the priority return, the `earlier` the processor is used
+in what's called the ``DataProcessorDelegate``. The delegate is a collection of registered ``DataProcessor`` of a
+particular ``DataManipulator`` such that all registered ``DataProcessor`` can be used.
+
+supports(DataHolder)
+--------------------
+
+Next up, we have ``supports(DataHolder)``. This is really simple, if your ``DataManipulator`` is supposed to be
+"supported" by the provided ``DataHolder`` (which can be an ItemStack, Entity, TileEntity, etc.) you can define the
+"support" in this method. So, if I have a ``VelocityDataProcessor``, my ``support(DataHolder)`` would look something like so:
```

Comment:
the `supports` doesn't handle serialization, that's for the `DataBuilder`, which isn't discussed yet.